src/adapter/z-stack/adapter/manager.ts

```diff
@@ -0,0 +1,445 @@
+                    this.debug.strategy("(stage-3) configuration-adapter mismatch (no backup)");
+                    return "startCommissioning";
+                }
+            }
+        }
+    }
+
+    /**
+     * Internal method to perform regular adapter startup in coordinator mode.
+     */
+    private async beginStartup(): Promise<void> {
+        const deviceInfo = await this.znp.request(Subsystem.UTIL, 'getDeviceInfo', {});
+        if (deviceInfo.payload.devicestate !== DevStates.ZB_COORD) {
+            this.debug.startup("starting adapter as coordinator");
+            const started = this.znp.waitFor(UnpiConstants.Type.AREQ, Subsystem.ZDO, 'stateChangeInd', {state: 9}, 60000);
+            await this.znp.request(Subsystem.ZDO, 'startupFromApp', {startdelay: 100}, null, [ZnpCommandStatus.SUCCESS, ZnpCommandStatus.FAILURE]);
+            await started.start().promise;
+            this.debug.startup("adapter successfully started in coordinator mode");
+        } else {
+            this.debug.startup("adapter is already running in coordinator mode");
+        }
+    }
+
+    /**
+     * Internal method to perform adapter restore.
+     */
+    private async beginRestore(): Promise<void> {
+        const backup = await this.backup.getStoredBackup();
+        /* istanbul ignore next */
+        if (!backup) {
+            throw Error("Cannot restore backup - none is available");
+        }
+
+        /* generate random provisioning network parameters */
+        const provisioningNwkOptions: Models.NetworkOptions = {
+            panId: 1 + Math.round(Math.random() * 65532),
+            extendedPanId: crypto.randomBytes(8),
+            channelList: [11 + Math.round((Math.random() * (29 - 11)))],
+            networkKey: crypto.randomBytes(16),
+            networkKeyDistribute: false
+        };
+
+        /* commission provisioning network */
+        await this.beginCommissioning(provisioningNwkOptions, false, false);
+        
+        /* perform NV restore */
+        await this.backup.restoreBackup(backup);
+
+        /* update commissioning NV items with desired nwk configuration */
+        await this.updateCommissioningNvItems(this.nwkOptions);
+        
+        /* settle & reset adapter */
+        await Wait(1000);
+        await this.resetAdapter();
+
+        /* startup with restored adapter */
+        await this.beginStartup();
+
+        /* write configuration flag */
+        await this.writeConfigurationFlag();
+    }
+
+    /**
+     * Internal method to perform new network commissioning. Network commissioning creates a new ZigBee
+     * network using the adapter.
+     * 
+     * @param nwkOptions Options to configure the new network with.
+     * @param failOnCollision Whether process should throw an error if PAN ID collision is detected.
+     * @param writeConfiguredFlag Whether Z2M `hasConfigured` flag should be written to NV.
```

Comment:
Why do you call this the Z2M hasConfigured flag, it applies to zigbee-herdsman right?

Reply:
Yes that's it. I just figured its Z2M specific. I may rename it to zigbee-herdsman flags id you wish 🙂

Reply:
Yes please

Reply:
Fixed